src/dev/beta.ts

```diff
@@ -0,0 +1,169 @@
+import { CancellationError } from '../shared/utilities/timeoutUtils'
+
+const localize = nls.loadMessageBundle()
+
+const downloadIntervalMs = 1000 * 60 * 60 * 24 // A day in milliseconds
+const betaToolkitKey = 'dev.beta'
+
+interface BetaToolkit {
+    readonly needUpdate: boolean
+    readonly lastCheck: number
+}
+
+function getBetaToolkitData(vsixUrl: string): BetaToolkit | undefined {
+    return globals.context.globalState.get<Record<string, BetaToolkit>>(betaToolkitKey, {})[vsixUrl]
+}
+
+async function updateBetaToolkitData(vsixUrl: string, data: BetaToolkit) {
+    await globals.context.globalState.update(betaToolkitKey, {
+        ...globals.context.globalState.get<Record<string, BetaToolkit>>(betaToolkitKey, {}),
+        [vsixUrl]: data,
+    })
+}
+
+/**
+ * Watch the beta VSIX daily for changes.
+ * If this is the first time we are watching the beta version or if its been 24 hours since it was last checked then try to prompt for update
+ */
+export function watchBetaVSIX(vsixUrl: string): vscode.Disposable {
+    const toolkit = getBetaToolkitData(vsixUrl)
+    if (!toolkit || toolkit.needUpdate || Date.now() - toolkit.lastCheck > downloadIntervalMs) {
+        runAutoUpdate(vsixUrl)
+    }
+
+    const interval = globals.clock.setInterval(() => runAutoUpdate(vsixUrl), downloadIntervalMs)
+    return { dispose: () => clearInterval(interval) }
+}
+
+async function runAutoUpdate(vsixUrl: string) {
+    getLogger().debug(`dev: checking ${vsixUrl} for a new version`)
+
+    try {
+        await telemetry.vscode_autoUpdateBeta.run(() => checkBetaUrl(vsixUrl))
+    } catch (e) {
+        if (!isUserCancelledError(e)) {
+            getLogger().warn(`dev: beta extension auto-update failed: %s`, e)
+        }
+    }
+}
+
+/**
+ * Prompt to update the beta extension when required
+ */
+async function checkBetaUrl(vsixUrl: string): Promise<void> {
+    const resp = await got(vsixUrl).buffer()
+    const latestBetaInfo = await getExtensionInfo(resp)
+    if (VSCODE_EXTENSION_ID.awstoolkit !== `${latestBetaInfo.publisher}.${latestBetaInfo.name}`) {
+        throw new ToolkitError('URL does not point to an AWS Toolkit artifact', { code: 'InvalidExtensionName' })
+    }
+
+    const currentVersion = vscode.extensions.getExtension(VSCODE_EXTENSION_ID.awstoolkit)?.packageJSON.version
+    if (latestBetaInfo.version !== currentVersion) {
+        const tmpFolder = await makeTemporaryToolkitFolder()
+        const betaPath = vscode.Uri.joinPath(vscode.Uri.file(tmpFolder), path.basename(vsixUrl))
+        await SystemUtilities.writeFile(betaPath, resp)
+
+        try {
+            await promptInstallToolkit(betaPath, latestBetaInfo.version, vsixUrl)
+        } finally {
+            await SystemUtilities.remove(tmpFolder)
+        }
+    } else {
+        await updateBetaToolkitData(vsixUrl, {
+            lastCheck: Date.now(),
+            needUpdate: false,
+        })
+    }
+}
+
+interface ExtensionInfo {
```

Comment:
I'm kind of surprised that this doesn't seem to exist in the vscode api :thinking:

Reply:
`packageJSON` exists on the extension interface but it's typed as `any`